pkg/gateway/controller.go

```diff
@@ -0,0 +1,256 @@
+				return
+			}
+			if exists {
+				c.gatewayCreatedOrUpdated(obj.(*unstructured.Unstructured))
+			}
+			c.queue.Forget(key)
+		}()
+	}
+}
+
+func (c *Controller) gatewayDeleted(obj interface{}, key string) {
+	var resource *unstructured.Unstructured
+	var ok bool
+	if resource, ok = obj.(*unstructured.Unstructured); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Could not convert object %v to DeletedFinalStateUnknown", obj)
+			return
+		}
+
+		resource, ok = tombstone.Obj.(*unstructured.Unstructured)
+		if !ok {
+			klog.Errorf("Could not convert object tombstone %v to Unstructured", tombstone.Obj)
+			return
+		}
+	}
+
+	haStatus, _, _ := getGatewayStatus(resource)
+	if haStatus == "active" {
+		c.clustersMap.Store(make(map[string]bool))
+	}
+}
+
+func (c *Controller) gatewayCreatedOrUpdated(obj *unstructured.Unstructured) {
+
+	haStatus, connections, ok := getGatewayStatus(obj)
+	if !ok || haStatus != "active" {
+		return
+	}
+	var newMap map[string]bool
+	currentMap := c.getClustersMap()
+	for _, connection := range connections {
+		connectionMap := connection.(map[string]interface{})
+
+		status, found, err := unstructured.NestedString(connectionMap, "status")
+		if err != nil || !found {
+			klog.Errorf("status field not found in %#v", connectionMap)
+		}
+		clusterId, found, err := unstructured.NestedString(connectionMap, "endpoint", "cluster_id")
+		if !found || err != nil {
+			klog.Errorf("clusterId field not found in %#v", connectionMap)
+			return
+		}
+
+		if status == "connected" {
+			_, found := currentMap[clusterId]
+			if !found {
+				if newMap == nil {
+					newMap = copyMap(currentMap)
+				}
+				newMap[clusterId] = true
+			}
+		} else {
+			_, found = currentMap[clusterId]
+			if found {
+				if newMap == nil {
+					newMap = copyMap(currentMap)
+				}
+				delete(newMap, clusterId)
+			}
+		}
+
+	}
+	if newMap != nil {
+		klog.Errorf("Updating the gateway status %#v", newMap)
+		c.clustersMap.Store(newMap)
+	}
+}
+
+func getGatewayStatus(obj *unstructured.Unstructured) (string, []interface{}, bool) {
+	status, found, err := unstructured.NestedMap(obj.Object, "status")
+	if !found || err != nil {
+		klog.Errorf("status field not found in %#v, err was: %v", obj, err)
+		return "", nil, false
+	}
+	haStatus, found, err := unstructured.NestedString(status, "haStatus")
+	if !found || err != nil {
+		klog.Errorf("haStatus field not found in %#v, err was: %v", status, err)
+		return "", nil, false
+	}
+	connections, found, err := unstructured.NestedSlice(status, "connections")
+	if !found || err != nil {
+		klog.Errorf("connections field not found in %#v, err was: %v", status, err)
+		return haStatus, nil, false
+	}
+	return haStatus, connections, true
+}
+
+func (c *Controller) getClustersMap() map[string]bool {
+	clustersMap := c.clustersMap.Load()
+	if clustersMap != nil {
+		return clustersMap.(map[string]bool)
+	}
+	return make(map[string]bool)
```

Comment:
The map should never be nil as it's initialized in the constructor. If it was then that would be a bug. Also returning a new map would hide the bug and result in the caller updating a local map. I suggest not checking and let it panic. ie

`return c.clustersMap.Load().(map[string]bool)`

Reply:
> The map should never be nil as it's initialized in the constructor. If it was then that would be a bug. Also returning a new map would hide the bug and result in the caller updating a local map. I suggest not checking and let it panic. ie
> 
> `return c.clustersMap.Load().(map[string]bool)`

What about in case of UT?

Reply:
Shouldn't UT call `NewController()`?